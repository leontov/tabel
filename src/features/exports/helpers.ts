import { saveAs } from 'file-saver';

export type CsvHeaders = Record<string, string>;
export type CsvRow = Record<string, string | number | undefined | null>;

export function buildCsv(headers: CsvHeaders, rows: CsvRow[]): string {
  const headerKeys = Object.keys(headers);
  const headerLine = headerKeys.map((key) => escapeValue(headers[key])).join(';');
  const body = rows
    .map((row) => headerKeys.map((key) => escapeValue(row[key])).join(';'))
    .join('\n');
  return [headerLine, body].filter(Boolean).join('\n');
}

function escapeValue(value: string | number | undefined | null): string {
  if (value === undefined || value === null) {
    return '';
  }
  const stringValue = String(value);
  if (stringValue.includes(';') || stringValue.includes('\"') || stringValue.includes('\n')) {
    return `"${stringValue.replace(/"/g, '""')}"`;
  }
  return stringValue;
}

export function downloadCsv(filename: string, headers: CsvHeaders, rows: CsvRow[]): void {
  const csv = buildCsv(headers, rows);
  const blob = new Blob([csv], { type: 'text/csv;charset=utf-8;' });
  saveAs(blob, filename);
}

export function downloadJson(filename: string, data: unknown): void {
  const blob = new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
  saveAs(blob, filename);
}
