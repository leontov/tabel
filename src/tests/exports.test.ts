import { describe, expect, it } from 'vitest';
import { buildCsv } from '../features/exports/helpers';

describe('CSV export', () => {
  it('creates localized CSV rows', () => {
    const csv = buildCsv(
      {
        person: 'Сотрудник',
        hours: 'Часы',
        amount: 'Сумма'
      },
      [
        { person: 'Анна', hours: '8.0', amount: '4800.00' },
        { person: 'Иван', hours: '6.5', amount: '3900.00' }
      ]
    );
    expect(csv).toMatchSnapshot();
  });
});
