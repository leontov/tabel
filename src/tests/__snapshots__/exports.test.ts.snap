// Vitest Snapshot v1

exports[`CSV export creates localized CSV rows 1`] = `
Сотрудник;Часы;Сумма
Анна;8.0;4800.00
Иван;6.5;3900.00
`;
